File: 01/main.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"strconv"
	"time"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func part1(frequencies *[]int) {
	var result int = 0
	for i := 0; i < len(*frequencies); i++ {
		frequency := (*frequencies)[i]
		result += frequency
	}
	fmt.Printf("Part 1: %v\n", result)
}

func part2(frequencies *[]int) {
	done := false
	var result int = 0
	seenFrequencies := make(map[int]bool)
	for !done {
		for i := 0; i < len(*frequencies); i++ {
			frequency := (*frequencies)[i]

			result += frequency

			if _, ok := seenFrequencies[result]; ok {
				fmt.Printf("Part 2: %v\n", result)
				done = true
				break
			}

			seenFrequencies[result] = true
		}
	}
}

func main() {
	start := time.Now()

	dat, err := ioutil.ReadFile("input.txt")
	check(err)

	lines := bytes.Split(dat, []byte("\r\n"))
	frequencies := make([]int, len(lines))
	for i := 0; i < len(frequencies); i++ {
		val, _ := strconv.Atoi(string(lines[i]))
		frequencies[i] = val
	}

	part1(&frequencies)
	part2(&frequencies)

	elapsed := time.Since(start)
	fmt.Print("Execution time: " + elapsed.String())
}

File: 01/main.js
const fs = require('fs');

const inputString = fs.readFileSync('input.txt').toString();
const frequencies = inputString.split('\r\n');

function executePart1(){
  let resultingFreq = 0;
  frequencies.forEach(frequency => {
    resultingFreq += parseInt(frequency);
  });
  console.log(`Part 1: ${resultingFreq}`);  
}

function executePart2(){
  let reachedSameFrequencyTwice = false;
  let resultingFreq = 0;
  const seenFrequencies = new Set();
  while (!reachedSameFrequencyTwice){
    for (let i = 0; i < frequencies.length; i++){
      resultingFreq += parseInt(frequencies[i]);
  
      if (seenFrequencies.has(resultingFreq)){
        console.log(`Part 2: ${resultingFreq}`);
        reachedSameFrequencyTwice = true;
        break;
      }
  
      seenFrequencies.add(resultingFreq);
    }
  }
}

const start = new Date();
executePart1();
executePart2();
const elapsed = new Date() - start;
console.log(`Total time: ${elapsed}ms`);